improve Float/Rational relations

smichr opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->

Not marking as closed until this is stable

closes #4473
 does not change #11707 since `Float(1,3) == Float(1,10)` but the hashes are different
closes #11072

#### Brief description of what is fixed or changed

Comparisons between Float and Rational were tightened up so results are more precise, e.g. `Float('1/3') != S(1)/3`...but this is currently under discussion.

#### Other comments

Initially this PR enforced a strict comparison of Float/Rational in expressions and a loose enforcement when comparing Numbers. This identified places in the code that weren't keeping track of whether the result should be a Float or Rational as in the subs `(x**(4.*y)).subs(x**(2.0*y), z)` which gave `z**2` instead of `z**2.0`.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
- core
    - comparison of Rational and Floats is more accurate
    - subs involving Float exponents will now retain the float as in `(x**(4.*y)).subs(x**(2.0*y), z) -> z**2.0` instead of `z**2`.
    - `equals` corrected to not make decisions based on numbers that evaluate without precision; self consistency based on solving for surds and an attempt to find minpoly are now attempted for all numberical expressions
    - Number can be used to instantiate oo and nan from strings
    - more stringent checks on the use of underscores in numbers passed as strings to Float are now made
    - checks for malformed mpf values passed to Float are now made
    - `1/S(0)` now returns `zoo`
- geometry
    - `ellipse.is_tangent` is more careful about deciding if a line that intersects at two points is not tangent -- it is possible to calculate an intersection as two point that are symbolically different but actually represent the same point in which case the line is tangent
    - `Point.affine_rank` now uses an `iszerofunc` that assumes numbers less than 1e-12 are zero during rank calculation.
<!-- END RELEASE NOTES -->

smichr committed:
more careful Float/Rational comparisons

smichr committed:
refactor number.__rel__ code

An important change here is that when the Rational is
recast for comparison with the Float, the Float's
precision is used.

This always affects rationals that are inexact in
base-2 representation (i.e. don't have a denominator that
is a power of 2)

>>> round_trips = lambda n: Rational(Float(n)) == n

>>> round_trip(S('1/3'))
False

It only affects exact rationals with
a numerator with more bits than the requested precision.

>>> round_trips(S('1/2'))
True

>>> odd = lambda n: prod([2*i+1 for i in range(1, n)])
>>> round_trips(odd(15)/S(2))
True
>>> round_trips(odd(16)/S(2))
False

The moral of the story is that Float, like Integer, will
truncate Rationals:

>>> odd(10)
654729075
>>> Rational(Float(_/S(2)))
654729075/2

>>> odd(20)
319830986772877770815625
>>> Rational(Float(_/S(2)))
159915493386438876069888

asmeurer commented:
I'm not sure if this behavior makes sense

```py
>>> S(2.0) == S(2)
True
>>> x + S(2.0) == x + S(2)
False
```

Shouldn't we have both true or both false? I'm leaning toward both true, since Float(2.0) is exactly equal to Integer(2) as a rational number.

asmeurer commented:
I like the other changes here, though, especially the more stringent float comparison that makes `0.1 == Rational(1, 10)` give False. That's actually a big first step to fixing #11707. I would suggest splitting them off. It may also make the tests easier to fix. I'm guessing some are due to more stringent float comparison and some are due to disallowing floats to equal rationals inside of expressions.

Also even if we do keep the changes here as is, I don't think this completely fixes #11707. 

```
>>> Float("0.1") == Rational(Float("0.1"))
True
>>> hash(Float("0.1")) == hash(Rational(Float("0.1")))
False
```

We need the hash algorithm discussed in the issue, so that numerically equal numbers give the same hashes. Alternately Rational == Float should never be true, but I think that needs more discussion if we chose that route.

smichr commented:
working on fixing some errors

jksuom commented:
> Rational == Float should never be true

I am beginning to believe that this is how it should be. A Float, represented by a rational number (an integral multiple of a power of 2), denotes a real number in an interval determined by precision. (The endpoints of the interval are also rational numbers.) `Rational == Float` could be None if the rational number falls within that interval, otherwise it should be false.

oscarbenjamin commented:
The issue with Rational != Float is transitivity:
```
S(2) == 2 == 2.0 == S(2.0)
```
Perhaps a way to break that chain would be if 2.0 != S(2.0).

smichr commented:
> `Rational == Float` could be None

Doesn't `==` have to give True or False?

smichr commented on sympy/core/numbers.py:
```diff
@@ -1745,93 +1733,93 @@ def __eq__(self, other):
+            # a Rational is always in reduced form so will never be 2/4
+            # so we can just check equivalence of args
+            return self.p == other.p and self.q == other.q
+        if other.is_Float:
```
@jksuom , this is where all the Float/Rational comparison logic takes place. I think I experimented with comparing them in an approximate way and returning True if they differed by less than the range permitted by the precision.

jksuom replied:
I would start with a quick test to see if the denominator is a power of 2:

    if self.q & (self.q - 1):
        return False

(Then we also know that `integer_log(self.q, 2)` will return True.)

smichr replied:
beautiful! done.

smichr commented on sympy/core/tests/test_expr.py:
```diff
@@ -1671,8 +1671,9 @@ def test_round():
-    assert all(r == i for i in range(9984, 10049))
+    # compares at equal precision (but it knows that
+    # the answer is not odd)
+    assert all(r == i for i in range(9984, 10049, 2))
```
@jksuom, I now see that the way the comparison is made already handles the "interval nature" of the Float by doing the comparison of the `Float` with `Float(Rational)` which forces the Rational either into the same bin as the Float or not. For this reason, there may be many Rationals that will compare as being equal to a Float as this tests shows. Given this fuzziness, is it a good idea to allow for parity discrimination? Notice how the Float `r` compares equal to many *even* values but no *odd* values.

jksuom replied:
> there may be many Rationals that will compare as being equal to a Float

It would follow from transitivity of == that those Rationals would also be equal. However, I think that transitivity is not implemented in SymPy, so that may not be an issue. But still, that does not feel right. It would be interesting to see the consequences of Rational == Float being always False. None would probably behave similarly (unless forbidden somewhere) but it is harder to implement. Making it False seems to me analogous to `x == 2` being False in SymPy although `x` could take the value 2.

smichr commented:
> We need the hash algorithm discussed in the issue, so that numerically equal numbers give the same hashes.

I just implemented exact comparisons but Float can still equal Rational.

I can't see a way to implement "equal numbers => equal hash", however. `Float(.5, 1) == Float(.5, 2) == 1/2` but they each have a different hash. To give them the same hash would require 

1) leaving off the precision information for the Floats -- but then does that mean that cache will use in calculations whichever came first? and 
1) store fractions with power-of-2 denominators like mpfs with infinite precision.

This would imply that we can only do `Float(.5, 1) != Float(.5, 2)` and `Float != Rational`. And there would be some justification in that: numbers with different don't behave the same (and from a real-world perspective even "cost" more as the precision increases) and a Rational represents an ideal...so they really are 3 different things.

jksuom commented:
> store fractions with power-of-2 denominators like mpfs with infinite precision

It looks like that Python does something like this when identifying floats with (some) rationals. That makes arithmetic laws dependent on the type of the operands.
```
>>> a = 1<<53
>>> b = float(a)
>>> a == b
True
>>> a + 1 == b + 1
False
```
I personally would prefer preservation of mathematical properties (as far as possible) even if that would make == not imply equal hash.

smichr commented:
> imply equal hash

I see that the precision was added to the hashable content [in this commit](https://github.com/sympy/sympy/commit/769b0489ca020facb7d942a63076d2b4bafc8253). What happens if we take out that part of the hash besides? Would this mean that the unintended higher precision value might be used in a calculation? Decimal hashes different precisions to the same value so there must be a way around this issue. Maybe they don't cache their numbers?

oscarbenjamin commented:
> I personally would prefer preservation of mathematical properties (as far as possible) even if that would make == not imply equal hash.

We need `a==b implies hash(a)==hash(b)` if sympified objects are to work properly in sets and dicts. This is explained in the docs:
https://docs.python.org/3.3/reference/datamodel.html#object.__hash__
The implementation of hash table data structures depends on the contrapositive `hash(a)!=hash(b) implies a!=b` to function correctly. Not respecting that will cause obscure and hard to track down problems. This is the sole purpose of `__hash__` so there is no point in defining it any other way.

The other constraint on this is that currently SymPy allows free interoperation with Python's int/float types. That in turn means that SymPy's `__eq__` needs to be consistent with theirs. If we said that `S(2)!=2` and `S(2.0)!=2.0` then we could freely defined `__eq__` for SymPy objects to do something different from what Python does for int/float.

If SymPy's `__eq__` is consistent with Python's then we can just use `fractions.__hash__` for all rational numbers (including normal Floats etc):
https://github.com/python/cpython/blob/c664b342a47e4b4650706d07e3e40a295e3a4407/Lib/fractions.py#L548

Python's Decimal type does hash floats with different precision to the same value but that's necessary because they compare equal:
```
In [27]: from decimal import Decimal                                                                                              

In [28]: d1 = Decimal('3.0')                                                                                                      

In [29]: d2 = Decimal('3.00000000000000')                                                                                         

In [30]: d1                                                                                                                       
Out[30]: Decimal('3.0')

In [31]: d2                                                                                                                       
Out[31]: Decimal('3.00000000000000')

In [32]: d1 == d2                                                                                                                 
Out[32]: True
```
Perhaps the difference is that the decimal library attaches actual working precision for calculations not to the Decimal instance but to the context in place at the time of the calculation:
```
In [36]: Decimal('2') / Decimal('3.00000')                                                                                        
Out[36]: Decimal('0.6666666666666666666666666667')
```

jksuom commented:
>  if sympified objects are to work properly in sets and dicts

I think that this should apply to sympified objects that can be used *interchangeably* anywhere, like two Floats that have the same `_mpf_` and `_prec`. But I am not sure that it would make sense to allow replacing e.g. Integer(2) freely with Float(2) even though == would be true.

oscarbenjamin commented:
> > if sympified objects are to work properly in sets and dicts
> 
> I think that this should apply to sympified objects that can be used _interchangeably_ anywhere, like two Floats that have the same `_mpf_` and `_prec`. But I am not sure that it would make sense to allow replacing e.g. Integer(2) freely with Float(2) even though == would be true.

Internally dicts/sets use `__eq__` as well as `__hash__`. The precise implementation varies across Python versions and implementations (i.e. pypy but also significant changes in recent CPython releases). Breaking the hash invariant leads to undefined behaviour and is not a supported or reliable way of ensuring that objects that compare equal are distinguished by hash-based containers. For example what happens when there's a hash collision?

If we want objects to be treated as distinct keys in e.g. a dict then the correct way to do that is to store keys that compare unequal with `__eq__`. Making the hashes also unequal can speed up dict lookup but is not essential for correctness.

We could add an attribute/method `_as_tuple` or `_cache_key` to `Basic` that returns something like a `(func, args)` tuple with the args also recursively tuplified. That could be used as a key for a caches etc. since that would only compare equal when two objects really are the same kind of SymPy object e.g.
```
>>> S(1)._as_tuple
(Integer, 1)
>>> S(1.0).as_tuple
(Float, ((0, 1, 0, 1), 53))
>>> exp(2)._as_tuple
(exp, ((Integer, 2),))
```

asmeurer commented:
Reminder that I've written about what can happen when a == b doesn't imply hash(a) == hash(b) https://www.asmeurer.com/blog/posts/what-happens-when-you-mess-with-hashing-in-python/. The example 3 behavior can be seen with Rational and Float in master (see the example at the top of https://github.com/sympy/sympy/issues/11707).

asmeurer commented:
I disagree with the notion that floating point numbers should be considered as intervals. It's better to think of floating point numbers (at some given precision n) as a subset of the rational numbers. The particular numbers which constitute that set is determined by n and the semantics of mpf, but just consider that there is *some* subset of rational numbers which are also floating point numbers (ignore infinity and nan; they don't matter for this discussion). Any arithmetic of numeric function on floating point numbers should produce the value in the set that is closest to the true mathematical value. One may also consider the "identity" function that maps a rational number to its closest floating point equivalent, for instance, `1.1 -> 2476979795053773*2**-51` (for the default binary precision 53).

The distinction is important. It's not the floating point numbers themselves that are inexact, but rather the arithmetic and numeric functions. Every floating point number is an exact rational number. However, not every rational number is a valid floating point number (for a given precision n), and must be mapped to the closest approximation when encountered, either as direct input from the user (`1.1`), or as the result of a numeric or arithmetic operation. To quote http://mpmath.org/doc/current/technical.html:

> Contrary to popular belief, floating-point numbers do not come with an inherent “small uncertainty”, although floating-point arithmetic generally is inexact. Every binary floating-point number is an exact rational number. With arbitrary-precision integers used for the mantissa and exponent, floating-point numbers can be added, subtracted and multiplied exactly. In particular, integers and integer multiples of 1/2, 1/4, 1/8, 1/16, etc. can be represented, added and multiplied exactly in binary floating-point arithmetic.

smichr commented:
> I disagree with the notion that floating point numbers should be considered as intervals. It's better to think of floating point numbers (at some given precision n) as a subset of the rational numbers.

Yes, I think we agree. But for purposes of testing -- not at the `__eq__/hash` level, however -- it is useful to think of two numbers being as good as equal if the Rational falls within the limits of representation of the Float against which it is being compared. I understand that this is *not* what is under consideration.

I have added a commit where I attempt to get equivalent Float and Rational hashing to the same values. All exact base-2 fractions (Integer, Float and some Rational) will all store an mpf as their hashable content while other Rationals will store their (p,q) tuple.

jksuom commented:
> Every floating point number is an exact rational number.

I would rather say that a floating point number is *represented* by an exact rational number. I don't think that they could be the same because their arithmetic properties are different. Moreover, there is a whole series of SymPy Floats of different precision represented by a single rational number. We should be able to distinguish between them when precision is changed in the course of a computation. Can that be done using `__eq__`?

smichr commented:
> Can that be done using `__eq__`?

If we let `2 == 2.0` then, like Decimal, I think we have to let `Float(2, 15) == Float(2, 100)`.

smichr commented:
Despite the comment in `Basic.__eq__` I don't know how to set this up with `__hash__` or `_hashable_content` so `hash(Float(3)) == hash(Integer(3))`. I think I have the `_hash` values sorted out, however: `Rational(a/2**b)` stores an `mpf` while other `Rational`s store `(p, q)`; `Float(int)` stores an `int` else an `mpf`.

oscarbenjamin commented:
> > Every floating point number is an exact rational number.
> 
> I would rather say that a floating point number is _represented_ by an exact rational number. I don't think that they could be the same because their arithmetic properties are different. Moreover, there is a whole series of SymPy Floats of different precision represented by a single rational number. We should be able to distinguish between them when precision is changed in the course of a computation. Can that be done using `__eq__`?

I agree with what you are saying but there are a series of design decisions in Python and SymPy that lead to the conclusion that Float precision needs to be ignored by `__eq__` (and `__hash__`). From SymPy's perspective this stems from the decision to have S(2)==2 and S(2.0)==2.0.

The overarching conflict here is that SymPy is implemented not just on top of Python but also to be used from within Python and to interoperate seamlessly with Python's existing types. If we were building something on top of Python with users writing their code in another language then there would be no need to allow Python's int/float to be mixed into any operations with SymPy types and we would be free to define these things however we liked. We could define one == to be used from Python code that really does do structural equality and another == for users to use (in the user language) that does mathematical equality.

For caching in particular this can be solved by not using a SymPy object directly as the cache key. Likewise in other situations where we use == but want to distinguish Floats with different precision or Float from Integer etc we could define a method/function or use some other key to achieve whatever effect is wanted in comparing the objects.

oscarbenjamin commented:
> I think I have the _hash values sorted out, however: Rational(a/2**b) stores an mpf while other Rationals store (p, q); Float(int) stores an int else an mpf.

Would it not make sense to put this in `Number.__hash__` and use as_numer_denom?

smichr commented:
Float  `as_numer_denom()` self gives (self, 1).

More importantly, in order to match Python we have to use has(self.p) when self equals an int.

smichr committed:
Number can instantiate oo, nan

smichr committed:
Float is more careful about oo, nan, _

smichr committed:
fix malformed mpf test; special is exact

smichr committed:
full path import

smichr committed:
fix tests/code for new Float/Rational comparisons

asmeurer commented:
Precision does complicate things a bit. If we use my idea of only thinking about the numeric value of a float for equality, and not its arithmetical properties, then the precision should be ignored, since it only affects the arithmetic properties. `Float('2.0', 2)` and `Float('2.0', 15)` both represent the exact same rational number (the `_mpf_` in both cases is `(0, 1, 1, 1)`, corresponding to `1*2**1`). The difference is that the former is a member of a smaller subset of rationals, so the truncation after operations will be different, e.g., `2.0 + 0.01`. 

I do worry about the effect of ignoring type (`Rational` vs. `Float`) and precision on the cache. Will the cache swap out floats for equivalent rationals, or two equal floats with unequal precisions? If so, this may be a larger concern than keeping transitivity with Python numbers (I actually don't think transitivity is that important, although `2 != 2.0` will be unexpected by most users, and `S(2) != int(2)` will be even more unexpected). Storing the type along with the value in the cache could solve the issue for Rational/Float, and it's probably a good idea anyway. I think most SymPy classes would suffer the same issue when subclassed without modifying their args. But that doesn't take care of precision. I wish there were a clever way to incorporate the precision into the hash. But if we want to use the same has as Python, so that `a == b -> hash(a) == hash(b)` if `a` is a SymPy Float/Rational and `b` is a Python int/float, then we have to use the same numeric hash algorithm as Python. I don't see how we can incorporate the precision into that.

oscarbenjamin commented:
Storing the type along with the object won't help if the object is not a simple Float/Rational e.g. `cos(2*x)` vs `cos(2.0*x)`.

Loss of transitivity means that the contents of `set(items)` can have different numbers of elements depending on the order of items. In Python we have
```python
>>> {2, 2.0}
{2}
>>> {2.0, 2}
{2.0}
```
What happens is that `__eq__` defines an equivalence class. The set contains exactly one element representing each equivalence class which is simply the first one encountered. In building the above sets once 2 or 2.0 has been added the other is not since it is equal to something already in the set.

We can define an Integer class that breaks transitivity by only comparing equal to int and not float like so:
```python
class Integer(object):
    '''
    Integer class that compares equal to int but not float.
    '''
    def __init__(self, num):
        self.num = num

    def __repr__(self):
        return 'Integer(%d)' % (self.num,)

    def __eq__(self, other):
        if isinstance(other, Integer):
            return self.num == other.num
        elif isinstance(other, int):
            return self.num == other
        elif isinstance(other, float):
            return False
        return NotImplemented

    def __hash__(self):
        return hash(self.num)
```
Then if we create sets that mix this with ints/floats we get:
```python
>>> from tmp import Integer
>>> {2, 2.0, Integer(2)}
{2}
>>> {2.0, Integer(2), 2}
{2.0, Integer(2)}
```
Now the number of items in the set depends on the order of the inputs. This is because the `__eq__` we have defined does not properly define equivalence classes by not satisfying transitivity.

smichr commented:
The hashing algorithm for Decimal (which is said to give the same value as for float, but I didn't check) is given [here](https://www.quora.com/How-does-Pythons-built-in-hash-function-work):

```python
def __hash__(self):
        """x.__hash__() <==> hash(x)"""
        # Decimal integers must hash the same as the ints
        #
        # The hash of a nonspecial noninteger Decimal must depend only
        # on the value of that Decimal, and not on its representation.
        # For example: hash(Decimal('100E-1')) == hash(Decimal('10')).
        if self._is_special:
            if self._isnan():
                raise TypeError('Cannot hash a NaN value.')
            return hash(str(self))
        if not self:
            return 0
        if self._isinteger():
            op = _WorkRep(self.to_integral_value())
            # to make computation feasible for Decimals with large
            # exponent, we use the fact that hash(n) == hash(m) for
            # any two nonzero integers n and m such that (i) n and m
            # have the same sign, and (ii) n is congruent to m modulo
            # 2**64-1.  So we can replace hash((-1)**s*c*10**e) with
            # hash((-1)**s*c*pow(10, e, 2**64-1).
            return hash((-1)**op.sign*Page on Op*pow(10, op.exp, 2**64-1))
        # The value of a nonzero nonspecial Decimal instance is
        # faithfully represented by the triple consisting of its sign,
        # its adjusted exponent, and its coefficient with trailing
        # zeros removed.
        return hash((self._sign,
                     self._exp+len(self._int),
                     self._int.rstrip('0')))
```

jksuom commented:
It seems that this is the old implementation. The [new version](https://github.com/python/cpython/blob/ad4ed872415d00fcdfaa52a08108ec752b115000/Lib/_pydecimal.py#L939-L963) looks compatible with the new implementation for ints and floats (no more the hash of a triple (sign, exp, int)). There is also a C-version in Modules/_decimal/_decimal.c but that appears much harder to read.

jksuom commented:
Python Decimal is analogous to SymPy Float in that both have arbitrary precision. A fundamental difference (apart from different base) is that Decimal has no precision attribute. Instead, there is a global context that includes user adjustable precision. There is no problem with equality and hash as precision need not be considered. The arithmetic operations depend on the context. Maybe a context could also be implemented in SymPy core (and the precision of Floats deprecated).

smichr commented:
It doesn't look like Decimal tries to match hash with float (though it does with int):
```python
>>> hash(Decimal('12.57')),hash(12.57)
(1481763729, 1997161115)
>>> hash(Decimal('12.')),hash(12.)
(12, 12)
>>> Decimal('12.57')==12.57
False
```

jksuom commented:
float(12.57) cannot be represented exactly by a decimal fraction, Fraction has to be used (and Python 3 for the new hash).
```
>>> hash(Fraction(1257, 100))
2144433998568735387
>>> hash(Decimal('12.57'))
2144433998568735387
>>> Decimal('12.57') == Fraction(1257, 100)
True
```
Multiples of negative powers of 2 can be exactly represented by floats.
```
>>> Decimal('12.25') == 12.25
True
>>> Fraction(1225, 100) == 12.25
True
```

asmeurer commented:
The Python hash algorithm is described here https://docs.python.org/3/library/stdtypes.html#hashing-of-numeric-types. I don't know if it can be be implemented more efficiently for float, since the denominator is always a power of 2 (it seems the built-in `pow` doesn't accept negative exponents with the third argument). I also don't know if it is already implemented in mpmath, but if it is, we should use that (assuming we do decide on this algorithm).

smichr committed:
improve equals; existing test now passes

jksuom commented on sympy/core/numbers.py:
```diff
@@ -1280,7 +1303,11 @@ def __eq__(self, other):
+            # and Float(.1, 100) == Float(.1)
             return bool(mlib.mpf_eq(self._mpf_, other._mpf_))
+        if other.is_Rational:
+            return other.__eq__(self)
```
The current implementation of `Rational.__eq__(Float)` will have to be changed if we want to implement the idea that a Float is equal to exactly one Rational, namely the one that represents it.

smichr replied:
Please check the current -- I think I have it down to 1:1

jksuom replied:
Most rational numbers are not equal to any Float, only those whose denominator is a power of 2. The current implementation converts Rational `self` to mpf:
https://github.com/sympy/sympy/blob/b024c2aa17f6b8d473d893d82a9f7533c132380b/sympy/core/numbers.py#L1777-L1778
The denominator of (the rational number corresponding to) the converted mpf is then a power of 2. That will make it always equal to some Float, which is not desirable.

We should probably first check that `d = self.q` is not a power of 2 (maybe by something like `d & (d - 1) != 0`) and return False in that case. If it is a power of 2, then it possible to convert it to an mpf with almost no computation and that can be used in the comparison.

jksuom commented on sympy/core/numbers.py:
```diff
@@ -1294,73 +1321,60 @@ def __eq__(self, other):
+        elif other.is_Rational:
+            # if you don't use self._prec here, the Rational
+            # will be choppped to default precision
+            other = Float(other, self._prec)
```
Would it be possible to use the same method as above:

        if other.is_Rational and not other.is_Integer:
            self *= other.q
            other = _sympify(other.p)

Then the full precision of `other` would be preserved.

smichr replied:
This is not for equality here so I am just scaling them to the same precision. There was also a test involving a gargantuan number (`E**1000`) that is making me defensive in not wanting to increase in size when dealing with Floats.

jksuom commented on sympy/core/numbers.py:
```diff
@@ -1745,93 +1759,95 @@ def __eq__(self, other):
+                op = getattr(other, attr)
+            elif other.is_Rational:
+                self, other = self.p*other.q, self.q*other.p
+                other = _sympify(other)
```
Should both `self` and `other` be made Integers?

smichr replied:
`self` is already the SymPy class type so this would be redundant.

smichr commented:
@jksuom , thanks for looking this over. I'm glad I currently have a schedule that can permit this and that the weather outside has been inclining toward winter again! This has been a real slog. Several of the other changes are motivated by the subtle change in how things are compared -- or, perhaps more likely, in my ignorance about how fragile some of the current code really is. Your input is a great encouragement as always.

smichr commented:
So do we have one exception to the `Float != Rational` rule: `0.0 == 0`? If not, probably all `foo == 0` need to be reviewed and perhaps replaced with `foo.is_zero`. And then maybe `1/0.0` giving `oo` but `1/0` giving `zoo` makes sense?

jksuom commented:
As to the Float != Rational rule, it seems that no decision was made so far. I think that letting a Float be equal to its representing rational number n/2**k could be a viable alternative. Then some Floats with different precision would be equal but Python is able to handle Decimals ignoring precision.

In any case, I think that `0.0 == 0` could be true. One reason is that "zero is zero at any precision", i.e., there is only one floating point zero and its arithmetic properties are the same as those of S.Zero. That would make `1/0.0` equal to `zoo` which I think is a better choice than `oo` or `-oo`. Those are one-sided limits along the real line and there is no preferred direction in `1/0.0`.

smichr committed:
handle 0==0.0, 0/0, 1/0, etc...

jksuom commented on sympy/core/numbers.py:
```diff
@@ -1286,26 +1286,25 @@ def __int__(self):
-            # precision is ignored so Float(.1) == Float(.1, 100)
-            # and Float(.1, 100) == Float(.1)
-            return bool(mlib.mpf_eq(self._mpf_, other._mpf_))
+            # compare at the same precision
```
I would prefer the way this was above, comparison of unmodified mpf's. In particular, floats should not be truncated if we want them to compare as in Python. (And they should be sympified in `Float.__new__` by `mlib.from_float(num)`, with default precision and rounding. That would preserve mantissa and exponent.)

smichr replied:
I wrote it this way so `Float(1, 5) == Float(1) == Float(1, 50) == 1.0`. If we only want to preserve the last pair then all that would need to be done would be to compare always at `max(15, min(self._pre, other._prec))`...at least that's my first hunch.

jksuom replied:
I was assuming that `mpf_eq` would do the "right thing" and return True for mpf's that represent the same rational number even if the mpf's were not exactly the same. I'll have to look into that more closely.

jksuom replied:
I see now, `mpf_eq` is too simple-minded. The mpf's have to be normalised first. But that should be done in a lossless way, with sufficiently high precision (`prec >=self._mpf_.bc` for `self` and similarly for `other`) to avoid truncation.

smichr replied:
Remember that we are not changing the original numbers. 

I see that I should have said that I was trying to preserve `Float(0.1, 5) == Float(0.1) == Float(0.1, 50) == 0.1` If we don't want that then we will only have `Float(0.1) == 0.1`.

For reference,
```
>>> a, b = Decimal('0.100'), Decimal('0.1')
>>> a==b
True
>>> hash(a)==hash(b)
True
>>> set([a, b]) == set([a])
True
```

smichr replied:
BTW, all Floats are normalized in `Float._new` so if we want to keep pure comparisons (i.e. considering precision) then one can just compare the `._mpf_` directly.

jksuom replied:
> Float(0.1, 5) == Float(0.1) == Float(0.1, 50) == 0.1

I don't think that this would be necessary. The denominator of 0.1 (as a rational number) is not a power of 2. Therefore it can only be approximated by Floats. Different precisions (5, 50) give different approximations and they are not expected to be equal. The situation is not comparable to Decimal whose arguments represent rational numbers with denominator a power of 10. (There does not seem to exist an analogous "binary decimal" representation for numbers whose denominator is an exact power of 2.)

jksuom commented:
> We need the hash algorithm discussed in the issue, so that numerically equal numbers give the same hashes. Alternately Rational == Float should never be true, but I think that needs more discussion if we choose that route.

I think that a Python compatible hash algorithm could be easily implemented. (And it would not be necessary to have always Rational != Float.)

Integer hash could be simply `hash(p)`. It could also be coded explicitly on Python 3 but the native implementation is more efficient as it is in C. In addition, Python 2 implementation is different and harder to reproduce in python.

For Float hash we could use `mpf_hash`. It looks compatible with floats and with Fractions whose denominator is a power of 2 on Python 3. It may not be fully compatible with Fractions on Python 2, but that is going to be dropped anyway.

For Rational, `hash(Fraction(p, q))` looks like the best implementation. That should work for both Python 2 and 3. There is a minor overhead as Fraction tries to reduce the representation. When the support for Python 2 is dropped, it is possible to copy the (concise) implementation of `Fraction.__hash__` with no need for reduction.

smichr committed:
handle Int/BooleanAtom comparison

smichr committed:
use comp for numerical failing optics tests

smichr committed:
use comp for failing tests

smichr committed:
numerical test of concrete result

A numerical tests was formerly being done in `equals`; with that
removed, a numerical `iszerofunc` routine has to be used for the
rank test.

smichr commented:
ping @asmeurer , @jksuom . It looks like this is going to be passing now. branch name not withstanding, I really don't want to get into the hash changes. Do we have a consensus for the current changes?

smichr commented on sympy/core/basic.py:
```diff
@@ -1883,21 +1883,28 @@ def _aresame(a, b):
+    from .numbers import Number
     from .function import AppliedUndef, UndefinedFunction as UndefFunc
+    if isinstance(a, Number) and isinstance(b, Number):
+        return a == b and a.__class__ == b.__class__
```
```suggestion
        return a.__class__ == b.__class__ and a == b
```

asmeurer replied:
Why was this change made? I imagine testing class equality should be faster than instance equality in the unequal case.

smichr commented on sympy/core/expr.py:
```diff
@@ -3491,7 +3519,7 @@ def round(self, n=None):
                 return rv
             # use str or else it won't be a float
-            return Float(str(rv), digits_needed)
+            return Float(str(rv), dps)  # keep same precision
```
This lets `round(12345.0, -4)` be `Float(10000.0)` instead of `Float(10000, 2)`.

smichr commented on sympy/core/basic.py:
```diff
@@ -1883,21 +1883,28 @@ def _aresame(a, b):
+    if isinstance(a, Number) and isinstance(b, Number):
+        return a.__class__ == b.__class__ and a == b
     for i, j in zip_longest(preorder_traversal(a), preorder_traversal(b)):
         if i != j or type(i) != type(j):
```
```suggestion
        if type(i) != type(j) or i != j:
```

asmeurer commented:
I'll try to take a look if I get a chance. This touches a lot of core code, so it may be worth checking benchmarks to see if there are any regressions.

jksuom reviewed (commented):
I am not sure why this review window popped up after I added a comment to numbers.py, but perhaps it  is convenient to say here that I like these changes. Hash code can be handled later.

smichr committed:
tighten up Float/Relational comparison

The added test fails in master.

smichr commented on sympy/core/numbers.py:
```diff
@@ -1308,73 +1329,62 @@ def __eq__(self, other):
+        elif other.is_Rational:
+            # if you don't use self._prec here, the Rational
+            # will be choppped to default precision
+            other = Float(other, precision=self._prec)
```
If I make that change
```
s = self*other.q
other = Float(other.p)
return _sympify(bool(
            op(self._mpf_, other._mpf_)))
```
Then I lose this invariant:
```
    # Float vs Rational
    # the rational form is less than the floating representation
    # at the same precision
    assert [i for i in range(15, 50) if Rational(pi.n(i)) > pi.n(i)] == []
    # this should be the same if we reverse the relational
    assert [i for i in range(15, 50) if pi.n(i) < Rational(pi.n(i))] == []
```
(All values >= 15 appear in the list.)

smichr replied:
I have a new approach and a test that fails in the previous commit and in master.

jksuom replied:
> other = Float(other.p)

It seems that this may also lose information if the precision is not set high enough.

smichr replied:
Either way information can be lost: `self*other.q` will truncate `q` if it is high-precision. But this is not the approach that I used.

jksuom replied:
> s = self*other.q

We only need the mpf of the product. It looks like that could be `mlib.mpf_mul(self._mpf_, mlib.from_int(other.q))`. With default precision 0 there should be no truncation.

smichr replied:
OK, that passes the tests in evalf, numbers and relational. It ends up taking about 15% longer to finish the evalf test suite, however.

smichr committed:
get 0.0 from Float._new

smichr committed:
use iszerofunc for Point.affine_rank

smichr commented:
@oscarbenjamin , here is how things behave right now: hash stores precision, `__eq__` uses only mpf:
```
>>> set([Float(1),Float(1,3)])
{1.00, 1.00000000000000}
>>> Float(1)==Float(1,3)
True
>>> Float(.1)==Float(.1,3)
False
```

smichr committed:
use self*other.q <> other.p

smichr committed:
Revert "use self*other.q <> other.p"

This reverts commit 62c42c3e25aa241307ea1d55001c100d6503cc9f.

The evalf suite takes 15% longer with that change.

smichr committed:
use Float._new when possible

smichr commented:
> there are any regressions

There will be regressions in anything that depended on `equals` incorrect behavior: using a precisionless evaluation of a number to make a decision. I was just thinking that as a first line of defense we could make relationals reject evaluation of a Float that had `_prec=1` so `unproved_zero.n() < 1` would yield something like `1.e-124 < 0` or raise an error saying that it didn't evaluate. Of course one could get around this by doing `z.n().n()` unless we also disabled evaluation of a precisionless number, too.

smichr committed:
tidy up eval_is_(+/-)ive

As suggested by https://github.com/rpisarev in 11072
with modifications.

jksuom commented on sympy/core/expr.py:
```diff
@@ -822,19 +821,17 @@ def _eval_is_negative(self):
-                return bool(not i and n < 0)
-            elif n._prec == 1 and (not i or i._prec == 1) and \
+            if r._prec != 1 and i._prec != 1:
+                return bool(not i and op(n))
```
Should this be `op(n2)`?

smichr replied:
no
```suggestion
                return bool(not i and op(r))
```

smichr replied:
Good eye, @jksuom -- it was formerly `n` but I changed it to `r` to avoid confusion with `n2`.

smichr committed:
Update sympy/core/expr.py

jksuom commented on sympy/core/numbers.py:
```diff
@@ -1308,73 +1329,96 @@ def __eq__(self, other):
+                o = Float._new([fnone, fzero, fone][sign(other) + 1], 1, zero=False)
+            else:
+                s = self
+                o = Float._new(other._as_mpf_val(self._prec), self._prec, zero=False)
```
I feel uneasy about this conversion. It may lead to equal mpfs for unequal numbers.
```
>>> a = Float('2/3')  # self
>>> b = Rational(2, 3)  # other
>>> a._mpf_
(0, mpz(6004799503160661), -53L, 53L)
>>> b._as_mpf_val(a._prec)
(0, mpz(6004799503160661), -53L, 53L)
```

smichr replied:
Yes, that is why the check is made in the next line to see whether `Rational(o) != other`. If so, it means there was a loss in the conversion and the more accurate test will be made, otherwise the conversion to Float was lossless.

smichr replied:
```python
>>> a = Float('2/3')
>>> b=Rational(2,3)
>>> Rational(Float._new(b._as_mpf_val(a._prec), a._prec))
6004799503160661/9007199254740992
>>> Rational(Float._new(b._as_mpf_val(a._prec), a._prec))==b
False
```

jksuom replied:
I would expect `Rational(o) != other` for most rational numbers. Could we just pass to the more accurate test?

smichr replied:
OK, change has been made.

smichr commented on sympy/core/expr.py:
```diff
@@ -822,19 +821,17 @@ def _eval_is_negative(self):
 
-            n, i = self.evalf(2).as_real_imag()
-            if not i.is_Number or not n.is_Number:
+            r, i = n2.as_real_imag()
```
```suggestion
            r, i = self.evalf(2).as_real_imag()  # self.evalf(2) != n2 which used _eval_evalf
```

smichr committed:
skip check of Rational(Float(other))==other

smichr commented on sympy/core/tests/test_sympify.py:
```diff
@@ -161,11 +161,19 @@ def test_sympify_bool():
 
 
+@XFAIL
+def test_issue_16772():
```
add `ans = [Rational(3, 10), Rational(1, 5)]` and perhaps remove the XFAIL...this is a useful 'feature': if you want an iterable trversed with the flag you have to pass it as a list.

oscarbenjamin commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -265,7 +265,7 @@ def test_geometric_sums():
 
     result = Sum(0.25**n, (n, 1, oo)).doit()
-    assert result == S(1)/3
+    assert abs(result - 1.0/3) < 1e-11
```
Why not `assert result == 1/3.`?

smichr replied:
I don't know what the precision is although I think something like `assert result == Float(1/3.0, precision=result._prec)` would work.

smichr replied:
BTW, I did try `result == 1/3.0` first but that failed. I assume this is because the precision of the result was not exactly 53 (but didn't check).

smichr replied:
Also, `1/3` should be the same as `1/3.0` in the test file, shouldn't it? It only failed for Py 2.7 as you can see in [this run](https://travis-ci.org/sympy/sympy/jobs/533019219).

smichr replied:
(Or maybe that was my mistake to use 1/3 instead of `1/3.` if truedivision was not imported.

oscarbenjamin replied:
I assume that the result is calculated as `Float(0.25)/(1-Float(0.25))` which happens to work out the same as `1/3.` or `Float(1/3.)`.

smichr replied:
It's a truedivision situation: the test was formerly S(1)/3; I changed it to 1/3 (which passed locally because I was using Py 3) but failed under Travis Py 2.7. I changed it to `1/3.0` or similar and it works.

asmeurer replied:
The file ought to have `from __future__ import division` to avoid issues like this. With Python 2 support only hanging around for a few more months it might not be worth bothering with this, but we really should have made the standard `__future__` imports (`division` and `print_function`) required in the code quality checker.

oscarbenjamin replied:
I think it's best not to have from `__future__` import division. Actually at some point I removed that from all test_ files. This example is different but mostly it just hides sloppy test code that uses floats when it should be using rational.

asmeurer replied:
If that is the goal, we need a more robust way of testing for that, because it will be moot once we drop Python 2 support. Not to mention it won't guard against cases like `4/2`, and possibly other subtleties where the test wouldn't fail in Python 2 even with integer division. Should we have a code quality guard against int/int, requiring an explicit float/int if that is what is intended (like `1./2`)? 

The `__future__.division` import generally helps prevent bugs from only testing things in Python 3. As I said, it's not super important since we are dropping Python 2 soon anyway, and most code should be tested for real in the test suite, which ought to catch any such errors as well.

smichr committed:
Update test_sums_products.py

smichr committed:
watch for oo in Frel

S.NaN._prec == -1 so the redundant check was removed

smichr committed:
update comp

jksuom commented on sympy/core/numbers.py:
```diff
@@ -1308,73 +1329,98 @@ def __eq__(self, other):
-        if other.is_Number and other is not S.NaN:
+            return other
+        elif other.is_Rational:
+            from sympy.functions.elementary.complexes import sign
```
On my system, comparisons like `a < b` (with `a = Float('2/3'), b = Rational(2, 3)`) take over 8 ms. (Probably because logarithms are relatively slow.)
On the other hand, this code

        if other.is_Rational:
            smpf = mlib.mpf_mul(self._mpf_, mlib.from_int(other.q))
            ompf = mlib.from_int(other.p)
            return _sympify(bool(op(smpf, ompf)))

takes only 21 μs.

smichr replied:
I'm not sure why I saw the difference in run times. I implemented your suggestion and put an inline justifying why we are working with mpf multiplication instead of naive `self*other.q`.

jksuom commented on sympy/core/numbers.py:
```diff
@@ -1388,23 +1388,10 @@ def _Frel(self, other, op):
                 else:
                     # self and other have the same sign and other
```
It is probably not necessary to look at the signs as `mpf_cmp` also implements the shortcut of different signs.

smichr replied:
The `is_integer` does not look like it will meet the expectations of the codebase.

smichr committed:
use mlib.from_int

smichr committed:
leave Float instantiation alone

smichr committed:
update 16772 tests and add 16859 XFAIL

smichr committed:
Revert "tidy up eval_is_(+/-)ive"

This reverts commit 5f31f6741566fdac0e70d1b083bb45e8edc44e6a.

smichr commented:
> Rational(a/2**b) stores an mpf

Sorry, @oscarbenjamin , I meant to say `Float(s*m/2**-e)` stores an mpf - I was just saying that Floats store a fraction having a denominator that is a power of 2 as an mpf value.

oscarbenjamin commented on sympy/core/expr.py:
```diff
@@ -707,63 +708,90 @@ def equals(self, other, failing_expression=False):
+                    # to see if we are in the right ballpark or not and if so
+                    # *then* the simplification will be attempted.
+                    if s.is_Symbol:
+                        sol = list(solveset(diff, s))
```
Can s be a Symbol if `diff.is_number`?

Also list() will raise TypeError if the result is not a FiniteSet.

oscarbenjamin commented on sympy/core/numbers.py:
```diff
@@ -1064,33 +1116,31 @@ def __new__(cls, num, dps=None, prec=None, precision=None):
+                        assert num[1] >= 0
+                        assert all(type(i) in (long, int) for i in num)
+                    except AssertionError:
+                        raise ValueError('malformed mpf: %s' % num)
```
Why isn't this just an if statement?

smichr committed:
use only solve in equals

smichr committed:
Merge branch 'master' into float-hash

oscarbenjamin commented on sympy/core/tests/test_expr.py:
```diff
@@ -1652,15 +1652,15 @@ def test_random():
     from sympy.abc import x
 
-    assert Float('0.1249999').round(2) == 0.12
+    assert str(Float('0.1249999').round(2)) == '0.12'
```
Can we use comp rather than str here?

oscarbenjamin replied:
(and below)

smichr replied:
when testing round I want to make sure that exactly these digits are shown

oscarbenjamin commented on sympy/sets/tests/test_fancysets.py:
```diff
@@ -518,7 +518,9 @@ def test_ComplexRegion_contains():
-    assert 0.5 + 0.6*I in c3
+    #assert (0.5 + 6*I/10) in c3
+    assert (S.Half + .6*I) in c3
+    assert (S.Half + 6*I/10) in c3
```
What's happening here?

smichr replied:
Hmmm. I will have to check. I would expect 0.5 to have passed and 0.6 to have failed. Given the two that are passing I would expect the comment-out one to pass, too.

oscarbenjamin commented on sympy/solvers/ode.py:
```diff
@@ -4055,7 +4055,7 @@ def ode_nth_order_reducible(eq, func, order, match):
     >>> from sympy.abc import x
     >>> f = Function('f')
-    >>> eq = Eq(x*f(x).diff(x)**2 + f(x).diff(x, 2))
+    >>> eq = Eq(x*f(x).diff(x)**2 + f(x).diff(x, 2), 0)
```
How did this get missed before? This should have been generating a warning which should have been turned into an error by the test runner. Maybe warnings don't raise in the doctests...

smichr replied:
It has been raising errors in the travis tests. That's how I saw it.

oscarbenjamin commented:
+1 from me

smichr commented:
Thanks for looking things over. I will see if @asmeurer or @jksuom have any further comments.

smichr committed:
review edits and comp modifications

smichr commented:
`comp` is more functional for testing results -- if you enter what you see for a result (as docstring shows) as the 2nd arg of `comp` it should report the computed number as being equal to the first.

@oscarbenjamin , the `contains` test has been checked and the full-float tests positive for containment.

jksuom commented:
I left a comment on testing equality. Otherwise I am +1 for merging this.

smichr committed:
add quick check that self.q = 2**e

smichr committed:
raise ValueError for bad input to perfect_power

smichr commented:
Will merge when tests pass.

asmeurer commented:
Sorry I wasn't able to review this sooner. I'll try to take a look now.

asmeurer commented:
> 1/S(0) now returns zoo

Hasn't this already been the case for some time?

asmeurer commented:
Maybe it was supposed to be `1/S(0.)`?

asmeurer commented:
I did a scan through and I think everything is fine here. We still need to fix the hashing issue. 

My only concern would be to make sure that all the algorithms are sufficiently tested with floats that aren't the default precision. Historically, there have been a lot of bugs in SymPy from that, usually wrong result bugs. I did see some tests that set the precision, so maybe this is already done.
